File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    vector<pair<int, int>> dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int minMinute = 0, c = 0;
        queue<pair<int, int>> q; 
        for (int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 2){
                    q.push({i,j});
                }else if(grid[i][j] == 1)
                    c++;
            }
        }
        while(!q.empty() && c != 0){
            int sz = q.size();
            while(sz--){
                int x = q.front().first;
                int y = q.front().second;
                q.pop();
                for (pair<int, int> d : dir){
                    int new_x = x + d.first;
                    int new_y = y + d.second; 
                    if(new_x >= 0 && new_x < m && new_y >= 0 && new_y < n && grid[new_x][new_y] == 1){
                        c--;
                        grid[new_x][new_y] = 2;
                        q.push({new_x, new_y});
                    }
                }
            }
            minMinute++;
        }
        if(c==0)
            return minMinute;
        else
            return -1;
    }
};

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        // 0 => Empty, 1 => Fresh, 2 => Rotten
        int m = grid.length;
        int n = grid[0].length, e = 0;
        ArrayDeque<int[]> q = new ArrayDeque<>();
        int rotten = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 2){
                    q.add(new int[]{i, j});
                    rotten++;
                }
                if(grid[i][j] == 0) e++;
            }
        }
        int total_rotten = m * n - e;
        boolean[][] visited = new boolean[m][n];
        
        int[][] dir = {{1,0},{0,1},{-1,0},{0,-1}};
        int minutes = 0;
        if(rotten == total_rotten) return 0;
        while(!q.isEmpty() && rotten < total_rotten){
            int size = q.size();
            for(int k = 0; k < size; k++){
                int[] d = q.poll();
                int x = d[0], y = d[1];
                visited[x][y] = true;
                for(int i = 0; i < dir.length; i++){
                    int new_x = d[0] + dir[i][0];
                    int new_y = d[1] + dir[i][1];
                    if(new_x >= 0 && new_x < m && new_y >= 0 && new_y < n && !visited[new_x][new_y]){
                        if(grid[new_x][new_y] == 1){
                            q.add(new int[]{new_x, new_y});
                            grid[new_x][new_y] = 2;
                            rotten++;
                        }
                    }
                }
            }
            minutes++;  
            if(rotten == total_rotten) return minutes;         
        }
        return -1;
    }
}